### PR TITLE
Fix flaky test routing::max_num_peers_limit

### PR DESCRIPTION
**Pull request by saketh-are**
**TL;DR**: We fix the flakiness of this test by switching from automatically formed connections to manually formed ones.

As mentioned in #9408, this test is quite flaky. It seems to be a long-standing issue (see #5777 and [comment here](https://github.com/near/nearcore/blob/988039b5179e9492c671f4f41e854a2032bd1930/chain/network/src/peer_manager/tests/routing.rs#L855-L873)). 

**Expected flow of the test:**

1. Configure three nodes with `max_num_peers` = `ideal_connections_lo` = `ideal_connections_hi` = 2. Enable outbound connections.
2. Start the three nodes. Due to configured parameters, the nodes should automatically form all 3 connections `0--1`, `1--2`, and `0--2`.
3. Check that the nodes have the expected connections.
4. Start the fourth node with the same config. This node will attempt to connect to the other nodes to reach 2 connections. The other nodes should reject it, since they are at their connection limit.
5. Wait for each of the first three nodes to reject a connection attempt from the fourth node.
6. Finally, check that the fourth node has no connections.

**How it goes wrong:**

This failing build [times out on the last step](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2419). It waits forever to observe a state in which the fourth node has no connections, indicating that the fourth node somehow manages to form one.

Enabling outbound on all nodes is producing [a race condition](https://github.com/near/nearcore/blob/988039b5179e9492c671f4f41e854a2032bd1930/chain/network/src/peer_manager/connection/mod.rs#L376-L398) in which two nodes might both attempt to form an outbound connection with the other at the same time.

The race condition is not a problem in itself, but in this test we also have `max_num_peers` configured very low. We end up running afoul of [this constraint](https://github.com/near/nearcore/blob/769b063721507664a5323b4b9c53b8c7fe9c0e02/chain/network/src/peer_manager/network_state/mod.rs#L260) intended to prevent us from accidentally forming too many connections. When the race condition occurs, both the attempted connections in either directions may end up being rejected, and with unpredictable timing.

It leads to some pretty weird behavior. Even before [starting the fourth node](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2326), we see some events in which connections fail with messages like [ConnectionLimitExceeded](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2306), [AlreadyStartedConnecting](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2306), or [Dropping handshake (network at max capacity)](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2282), even though the first 3 nodes should be able to form all possible connections among themselves.

**The fix:**

Disabling `enable_outbound` and manually triggering the necessary connections (each in a single direction) allows the test to pass reliably. The test still verifies the fundamental behavior it is supposed to, which is the rejection of additional connections after `max_num_peers` is reached.

### CODE REVIEW

**Comment by walnut-the-cat** [comment]
Just random question. Why did you mention 'ideal_connections_lo' twice in the description? 

> Configure three nodes with max_num_peers = ideal_connections_lo = ideal_connections_lo = 2. Enable outbound connections.

Typo?

**Comment by saketh-are** [comment]
@walnut-the-cat just fixed, thanks, it's supposed to be `ideal_connections_lo` and `ideal_connections_hi`. In addition to having a hard limit `max_num_peers`, each node has a soft range set by `lo` and `hi` it tries to stay within by creating/dropping connections.

File: chain/network/src/peer_manager/tests/routing.rs [hunk 1]
```diff
@@ -787,7 +787,7 @@ async fn max_num_peers_limit() {
     let chain = Arc::new(data::Chain::make(&mut clock, rng, 10));
 
     tracing::info!(target:"test", "start three nodes with max_num_peers=2");
-    let mut cfgs = make_configs(&chain, rng, 4, 4, true);
+    let mut cfgs = make_configs(&chain, rng, 4, 4, false);
```
Comment (walnut-the-cat) [inline chain/network/src/peer_manager/tests/routing.rs:790]: what does the last boolean mean? turning off auto connection?

File: chain/network/src/peer_manager/tests/routing.rs [hunk 2]
```diff
@@ -802,6 +802,10 @@ async fn max_num_peers_limit() {
     let id1 = pm1.cfg.node_id();
     let id2 = pm2.cfg.node_id();
 
+    pm0.connect_to(&pm1.peer_info(), tcp::Tier::T2).await;
+    pm0.connect_to(&pm2.peer_info(), tcp::Tier::T2).await;
+    pm1.connect_to(&pm2.peer_info(), tcp::Tier::T2).await;
+
```
Comment (wacban) [inline chain/network/src/peer_manager/tests/routing.rs:808]: very ignorable nit: 0->1, 1->2, 2->0

**Comment by saketh-are** [comment]
> Why would we see error message, [ConnectionLimitExceeded](https://buildkite.com/nearprotocol/nearcore/builds/29836#0189dc51-e14c-4be6-aeaf-662f3de1ccf3/93-2306), when total number of connection stays below the limit? Does the count include pending connections that are started but not established as well?

Yep, [this condition linked here](https://github.com/near/nearcore/blob/769b063721507664a5323b4b9c53b8c7fe9c0e02/chain/network/src/peer_manager/network_state/mod.rs#L260) takes into consideration established connections + currently attempted outbound connections + attempted inbound connection. And if it's violated it leads to [ConnectionLimitExceeded](https://github.com/near/nearcore/blob/769b063721507664a5323b4b9c53b8c7fe9c0e02/chain/network/src/peer_manager/network_state/mod.rs#L327).

So in this test we might see something like: `0<-->1` is established successfully, then `0-->2` is attempted at the same time as `2-->0`.

**Review by wacban — APPROVED** [review]
lgtm, thanks for fixing this!

File: chain/network/src/peer_manager/tests/routing.rs (hunk 1 above)
Comment (saketh-are) [reply chain/network/src/peer_manager/tests/routing.rs:790]: That's right, see here:
https://github.com/near/nearcore/blob/769b063721507664a5323b4b9c53b8c7fe9c0e02/chain/network/src/peer_manager/tests/routing.rs#L634-L640

**Comment by saketh-are** [comment]
> The fix itself makes sense, but I now wonder how the we handle aforementioned race condition in real like scenario.

The way this test was written is overly constrained:
- lo=hi=max is not a config we'd see in the real world, it is just used to construct the specific test.
- The test uses auto-connect behavior but expects specific connections to be formed and maintained.

This would not occur in a real scenario with lo < hi < max. And even if it did, it would not cause any issues since nodes do not assume that auto-connect provides them with connections to specific other nodes.

The important thing for the real use case is that we prevent duplicate connections from forming if two nodes try to connect to each other at the same time, which we do.

I don't think any changes are warranted outside of cleaning up the test behavior.
